File: src/impls/ruby/polish.rb
#!/usr/bin/env ruby
# SPDX-FileCopyrightText: 2022 smdn <smdn@smdn.jp>
# SPDX-License-Identifier: MIT
class Node
  @expression
  @left
  @right

  attr_reader :expression

  def initialize(expression)
    @expression = expression
  end

  def parse
    pos_operator = get_operator_pos(expression)

    if pos_operator < 0
      @left = nil
      @right = nil
      return
    end

    # left-hand side
    @left = Node.new(remove_bracket(@expression[0, pos_operator]))
    @left.parse

    # right-hand side
    @right = Node.new(remove_bracket(@expression[pos_operator + 1, @expression.length]))
    @right.parse

    # operator
    @expression = @expression[pos_operator, 1]
  end

  def remove_bracket(s)
    return s unless s =~ /^\((.*)\)$/

    removed = $1.dup

    nest = 1

    removed.length.times do |i|
      if removed[i, 1] == '('
        nest += 1
      elsif removed[i, 1] == ')'
        nest -= 1
      end

      return s if nest == 0
    end

    raise("unbalanced bracket: #{s}") unless nest == 1

    if removed[0, 1] == '('
      return remove_bracket(removed)
    else
      return removed
    end
  end

  def get_operator_pos(exp)
    return -1 if exp.to_s.empty?

    pos = -1
    nest = 0
    priority = 0
    lowest_priority = 4

    exp.length.times do |i|
      case exp[i, 1]
        when '=' then priority = 1
        when '+' then priority = 2
        when '-' then priority = 2
        when '*' then priority = 3
        when '/' then priority = 3
        when '(' then nest += 1; next
        when ')' then nest -= 1; next
        else next
      end

      if nest == 0 and priority <= lowest_priority
        lowest_priority = priority
        pos = i
      end
    end

    return pos
  end

  def traverse_postorder
    @left .traverse_postorder if @left
    @right.traverse_postorder if @right

    print @expression
  end

  def traverse_inorder
    print '(' if @left and @right

    @left .traverse_inorder if @left

    print @expression

    @right.traverse_inorder if @right

    print ')' if @left and @right
  end

  def traverse_preorder
    print @expression

    @left .traverse_preorder if @left
    @right.traverse_preorder if @right
  end

  def calculate
    if @left and @right
      left_operand  = @left.calculate()
      right_operand = @right.calculate()

      case @expression
        when '+' then return left_operand + right_operand
        when '-' then return left_operand - right_operand
        when '*' then return left_operand * right_operand
        when '/' then return left_operand / right_operand
        else return 0.0
      end
    else
      return @expression.to_f
    end
  end
end

print 'input expression: '

root = Node.new(gets.chomp.gsub(' ', ''))

print "expression: #{root.expression}\n"

root.parse

print "reverse polish notation: "
root.traverse_postorder
print "\n"

print "infix notation: "
root.traverse_inorder
print "\n"

print "polish notation: "
root.traverse_preorder
print "\n"

print "calculated result: "
print root.calculate
print "\n"

File: src/impls/javascript/polish.js
#!/usr/bin/env node
// SPDX-FileCopyrightText: 2022 smdn <smdn@smdn.jp>
// SPDX-License-Identifier: MIT
import * as readline from "readline";
import { Node, MalformedExpressionError } from "./Node.mjs";

let rl = readline.createInterface({
  input: process.stdin,
  terminal: false,
});

if (rl) {
  process.stdout.write("input expression: ");

  process.exitCode = 1; // 入力がなかった場合は終了コード1で終了させる

  // 標準入力から二分木に分割したい式を入力する
  rl.on("line", expression => {
    process.exitCode = 0
    main(expression);
    rl.close();
  });
}

// メインの処理メソッド。　結果によって次の終了コードを設定する。
//   0: 正常終了 (二分木への分割、および式全体の値の計算に成功した場合)
//   1: 入力のエラーによる終了 (二分木への分割に失敗した場合)
//   2: 計算のエラーによる終了 (式全体の値の計算に失敗した場合)
function main(_expression) {
  if (!_expression)
    // 入力が得られなかった場合は、処理を終了する
    process.exit(1);

  // 入力された式から空白を除去する(空白を空の文字列に置き換える)
  let expression = _expression.replaceAll(" ", "");

  if (expression.length == 0)
    // 空白を除去した結果、空の文字列となった場合は、処理を終了する
    process.exit(1);

  let root = null

  try {
    // 二分木の根(root)ノードを作成し、式全体を格納する
    root = new Node(expression);

    console.log("expression: " + expression);

    // 根ノードに格納した式を二分木へと分割する
    root.parseExpression();
  }
  catch (err) {
    if (err instanceof MalformedExpressionError) {
      console.error(err.message);
      process.exit(1);
    }
    else {
      throw err;
    }
  }

  // 分割した二分木を帰りがけ順で巡回して表示する(前置記法/逆ポーランド記法で表示される)
  process.stdout.write("reverse polish notation: ");
  root.writePostorder(process.stdout);
  process.stdout.write("\n");

  // 分割した二分木を通りがけ順で巡回して表示する(中置記法で表示される)
  process.stdout.write("infix notation: ");
  root.writeInorder(process.stdout);
  process.stdout.write("\n");

  // 分割した二分木を行きがけ順で巡回して表示する(後置記法/ポーランド記法で表示される)
  process.stdout.write("polish notation: ");
  root.writePreorder(process.stdout);
  process.stdout.write("\n");

  // 分割した二分木から式全体の値を計算する
  let resultValue = root.calculateExpressionTree()

  if (resultValue !== undefined) {
    // 計算できた場合はその値を表示する
    console.log("calculated result: " + Node.formatNumber(resultValue));
  }
  else {
    // (式の一部あるいは全部が)計算できなかった場合は、計算結果の式を中置記法で表示する
    process.stdout.write("calculated expression: ");
    root.writeInorder(process.stdout);
    process.stdout.write("\n");
    process.exit(2);
  }
}
